internal/controller/nnf_node_storage_controller.go

```diff
@@ -256,50 +270,73 @@ func (r *NnfNodeStorageReconciler) deleteAllocation(ctx context.Context, nnfNode
 	return nil, nil
 }
 
-func (r *NnfNodeStorageReconciler) createAllocation(ctx context.Context, nnfNodeStorage *nnfv1alpha1.NnfNodeStorage, index int) (*ctrl.Result, error) {
-	log := r.Log.WithValues("NnfNodeStorage", client.ObjectKeyFromObject(nnfNodeStorage), "index", index)
+func (r *NnfNodeStorageReconciler) createAllocations(ctx context.Context, nnfNodeStorage *nnfv1alpha1.NnfNodeStorage, blockDevices []blockdevice.BlockDevice, fileSystems []filesystem.FileSystem) (*ctrl.Result, error) {
+	log := r.Log.WithValues("NnfNodeStorage", client.ObjectKeyFromObject(nnfNodeStorage))
 
-	blockDevice, fileSystem, err := getBlockDeviceAndFileSystem(ctx, r.Client, nnfNodeStorage, index, log)
-	if err != nil {
-		return nil, err
-	}
+	for index, blockDevice := range blockDevices {
+		allocationStatus := &nnfNodeStorage.Status.Allocations[index]
 
-	allocationStatus := &nnfNodeStorage.Status.Allocations[index]
-	ran, err := blockDevice.Create(ctx, allocationStatus.Ready)
-	if err != nil {
-		return nil, dwsv1alpha2.NewResourceError("could not create block devices").WithError(err).WithMajor()
-	}
-	if ran {
-		log.Info("Created block device", "allocation", index)
-	}
+		// Skip allocations that are already created
+		if allocationStatus.Ready {
+			continue
+		}
 
-	// We don't need to activate the block device here. It will be activated either when there is a mkfs, or when it's used
-	// by a ClientMount
-	ran, err = fileSystem.Create(ctx, allocationStatus.Ready)
-	if err != nil {
-		return nil, dwsv1alpha2.NewResourceError("could not create file system").WithError(err).WithMajor()
-	}
-	if ran {
-		log.Info("Created file system", "allocation", index)
-	}
+		ran, err := blockDevice.Create(ctx, allocationStatus.Ready)
+		if err != nil {
+			return nil, dwsv1alpha2.NewResourceError("could not create block devices").WithError(err).WithMajor()
+		}
+		if ran {
+			log.Info("Created block device", "allocation", index)
+		}
 
-	ran, err = fileSystem.Activate(ctx, allocationStatus.Ready)
-	if err != nil {
-		return nil, dwsv1alpha2.NewResourceError("could not activate file system").WithError(err).WithMajor()
-	}
-	if ran {
-		log.Info("Activated file system", "allocation", index)
-	}
+		_, err = blockDevice.Activate(ctx)
+		if err != nil {
+			return nil, dwsv1alpha2.NewResourceError("could not activate block devices").WithError(err).WithMajor()
+		}
 
-	ran, err = fileSystem.SetPermissions(ctx, nnfNodeStorage.Spec.UserID, nnfNodeStorage.Spec.GroupID, allocationStatus.Ready)
-	if err != nil {
-		return nil, dwsv1alpha2.NewResourceError("could not set file system permissions").WithError(err).WithMajor()
-	}
-	if ran {
-		log.Info("Set file system permission", "allocation", index)
+		deferIndex := index
+		defer func() {
+			_, err = blockDevices[deferIndex].Deactivate(ctx, false)
+			if err != nil {
+				allocationStatus.Ready = false
+			}
+		}()
 	}
 
-	allocationStatus.Ready = true
+	for index, fileSystem := range fileSystems {
```

Comment:
Won't this loop be running at the same time as the defer funcs from above are running? Touching the same fields in nnfNodeStorage.Status.Allocations[]?

Reply:
The defer funcs don't get run until we return from the CreateAllocations() function. So this loop happens before that.

Reply:
Oh, the surrounding function, not the surrounding block. Got it.